server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -136,34 +171,61 @@ protected void newResponseAsync(
             clusterService.threadPool().absoluteTimeInMillis()
         );
 
+        var remoteClusterStatsListener = ((ActionListenerWithRemotes) listener).getRemoteClusterStats();
+
         final ListenableFuture<MappingStats> mappingStatsStep = new ListenableFuture<>();
         final ListenableFuture<AnalysisStats> analysisStatsStep = new ListenableFuture<>();
         mappingStatsCache.get(metadata, cancellableTask::isCancelled, mappingStatsStep);
         analysisStatsCache.get(metadata, cancellableTask::isCancelled, analysisStatsStep);
+
         mappingStatsStep.addListener(
             listener.delegateFailureAndWrap(
                 (l, mappingStats) -> analysisStatsStep.addListener(
                     l.delegateFailureAndWrap(
-                        (ll, analysisStats) -> ActionListener.completeWith(
-                            ll,
-                            () -> new ClusterStatsResponse(
-                                System.currentTimeMillis(),
-                                metadata.clusterUUID(),
-                                clusterService.getClusterName(),
-                                responses,
-                                failures,
-                                mappingStats,
-                                analysisStats,
-                                VersionStats.of(metadata, responses),
-                                clusterSnapshotStats
+                        (ll, analysisStats) -> remoteClusterStatsListener.<ClusterStatsResponse>andThen(
+                            (delegate, remoteClusterStats) -> delegate.onResponse(
+                                new ClusterStatsResponse(
+                                    System.currentTimeMillis(),
+                                    metadata.clusterUUID(),
+                                    clusterService.getClusterName(),
+                                    responses,
+                                    failures,
+                                    mappingStats,
+                                    analysisStats,
+                                    VersionStats.of(metadata, responses),
```

Comment:
I'm concerned that on cancellation this will retain the (potentially quite large) `responses` collection while waiting for the `remoteClusterStatsListener` to complete, even though it's going to ultimately fail. Could we take steps to release this sooner? That's kind of another instance of the bug we fixed in https://github.com/elastic/elasticsearch/issues/55550.

Reply:
We could add a check in `newResponseAsync` and short-circuit it somehow in that case?